src/graphql/directive/mongoose_graphql_directive_use.erl

```diff
@@ -0,0 +1,135 @@
+%% @doc The custom directive `@use' specifies which modules or services have to be loaded
+%% to execute the command. We can annotate both objects and fields. The args from object
+%% annotation are aggregated and checked for each annotated object's field. Thus annotating
+%% only a category is not enough because, on the object level, we do not know the host type
+%% needed to check loaded modules.
+%%
+%% In below example <i>command1</i> will be checked for a loaded modules, but <i>command2</i>
+%% will be not because is not annotated. The admin endpoint does not have a host type in context,
+%% so we need to specify the `arg'.
+%% ```
+%% type Category @use(modules: ["module_a"]){
+%%     command1(domain: String!): String @use(arg: "domain")
+%%     command2: String
+%%}
+%%'''
+%%
+%% The user's endpoint context contains the authenticated user, so the host type is there,
+%% and we do not need to specify the `arg'.
+%% ```
+%% type Category @use(modules: ["module_a"]){
+%%     command1: String @use
+%%     command2: String
+%%}
+%%'''
+
+-module(mongoose_graphql_directive_use).
+
+-behaviour(mongoose_graphql_directive).
+
+-export([handle_directive/3, handle_object_directive/3]).
+
+-include_lib("graphql/src/graphql_schema.hrl").
+-include_lib("graphql/include/graphql.hrl").
+-include_lib("jid/include/jid.hrl").
+
+-include("mongoose.hrl").
+
+-import(mongoose_graphql_directive_helper, [name/1, get_arg/2]).
+
+-type host_type() :: mongooseim:host_type().
+-type ctx() :: mongoose_graphql_directive:ctx().
+-type use_ctx() ::
+    #{modules := [binary()],
+      services := [binary()],
+      arg => binary(),
+      atom => term()}.
+
+%% @doc Check the collected modules and services and swap the field resolver if any of them
+%% is not loaded. The new field resolver returns the error that some modules or services
+%% are not loaded.
+handle_directive(#directive{id = <<"use">>, args = Args}, #schema_field{} = Field, Ctx) ->
+    #{modules := Modules, services := Services} = UseCtx = aggregate_use_ctx(Args, Ctx),
+    ArgValue = get_arg_value(UseCtx, Ctx),
+    % Assume that loaded modules can be checked only when host type can be obtained
+    case host_type_from_arg(ArgValue) of
+        {ok, HostType} ->
+            UnloadedModules = filter_unloaded_modules(HostType, Modules),
+            UnloadedServices = filter_unloaded_services(Services),
+            case {UnloadedModules, UnloadedServices} of
+                {[], []} ->
+                    Field;
+                {_, _} ->
+                    Fun = resolve_not_loaded_fun(UnloadedModules, UnloadedServices),
+                    Field#schema_field{resolve = Fun}
+            end;
+        {error, not_found} ->
+            Field
```

Comment:
Wouldn't it be better to return something like "Invalid module or host type"? For me it looks counter-intuitive that we skip the check as if everything was fine.

Reply:
I was thinking about it and decided not to return an error here because we check domain/host type again in the API modules. The error returned from API modules is more precise because we know there if the domain or host type is nonexistent.

If we return an error here, then we stop testing the error returned from API modules because we override it. If this is not a problem, then I can return an error here.

Reply:
We have to create a follow-up ticket to add the `use` directive to other modules. I think we can add returning an error there if we decide to do it. What do you think?

Reply:
It is ok for now